File: example_code/bit_bucket.py
# Copyright (c) 2012 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.
from common_utils import FormatAsBits
from common_utils import StrToList
import sys
import struct

class BitBucket:
  """
  This class allows for bit-level manipulations of a list of bits.
  In particular, it allows for the storage of bits (or sets of bits), and
  it allows for the fetching of those stored bits (effectively in a FIFO manner)
  """
  def __init__(self):
    self.Clear()

  def Clear(self):
    """
    Clears out all data and resets the BitBucket to like-new state
    """
    self.output = []
    self.out_byte = 0
    self.out_boff = 0
    self.idx_byte = 0
    self.idx_boff = 0

  def AdvanceToByteBoundary(self):
    """
    Inserts enough '0's to ensure that the number of bits stored % 8 == 0
    """
    bits_to_advance = (8 - self.idx_boff) % 8
    if bits_to_advance:
      self.idx_boff += bits_to_advance
      self.idx_boff %= 8
      self.idx_byte += 1

  def StoreBit(self, bit):
    """
    Stores a single bit.
    """
    if bit:
      bit = 1
    else:
      bit = 0
    self.StoreBits( ([bit << 7], 1) )

  def StoreBits8(self, val):
    """
    Stores the 8 least-significant-bits from val
    """
    tmp_val = struct.pack(">B", val)
    self.StoreBits( (StrToList(tmp_val), 8))

  def StoreBits16(self, val):
    """
    Stores the 16 least-significant-bits from val in network order (big endian)
    """
    tmp_val = struct.pack(">H", val)
    self.StoreBits( (StrToList(tmp_val), 16))

  def StoreBits32(self, val):
    """
    Stores the 32 least-significant-bits from val in network order (big endian)
    """
    tmp_val = struct.pack(">L", val)
    self.StoreBits( (StrToList(tmp_val), 32))

  def StoreBits(self, input_tuple):
    """
    (inp_bytes, inp_bits) = input_tuple
    Stores inp_bits from inp_bytes. When inp_bits < len(inp_bytes)*8, the
    most-significant-bits (this is opposite the other StoreBits) of the last
    element of inp_bytes are used.
    """
    (inp_bytes, inp_bits) = input_tuple
    old_out_boff = self.out_boff
    if not inp_bytes:
      return
    if inp_bits % 8:
      leftover_bits = inp_bits % 8
    else:
      leftover_bits = 8
    if self.out_boff == 0:
      self.output.extend(inp_bytes)
      if not type(inp_bytes[0]) == int:
        print "type(inp_bytes[0]) == ", type(inp_bytes[0])
        print repr(input)
        raise StandardError()
      self.output[-1] &= ~(255 >> leftover_bits)
      self.out_boff = leftover_bits % 8
    else:
      # We know there is a non-zero bit offset if we're below here.
      # This also implies there MUST be a byte in output already.
      bits_left_in_byte = 8 - self.out_boff
      for c in inp_bytes:
        self.output[-1] |= c >> self.out_boff
        self.output.append(0)
        self.output[-1] = (c << bits_left_in_byte) & 255
      c = inp_bytes[-1]
      if self.out_boff + leftover_bits <= 8:
        self.output.pop()
        c = inp_bytes[-1]
        self.output[-1] |= c >> self.out_boff
      self.out_boff = (self.out_boff + leftover_bits) % 8
      if self.out_boff != 0:
        self.output[-1] &= ~(255 >> self.out_boff)
    if self.out_boff != (old_out_boff + inp_bits) % 8:
      raise StandardError()

  def GetAllBits(self):
    """ Returns a tuple containing (list-of-bytes, number-of-bits)
    When number-of-bits % 8 != 0, the last byte in list-of-bytes
    will have the remaining bits (number-of-bits % 8) stored
    from the most-significant bit onward towards the least-significant bit
    """
    return (self.output, self.NumBits())

  def NumBits(self):
    """
    Returns the number of bits stored into this BitBucket
    """
    num_bits = 8*len(self.output)
    if self.out_boff % 8:
      num_bits -= 8
      num_bits += self.out_boff
    if num_bits < 0:
      print "What the..."
    return num_bits

  def BytesOfStorage(self):
    """
    Returns the number of bytes necessary to hold all of the bits which have
    been stored into this BitBucket
    """
    return (self.NumBits() + 7) / 8

  def BitsRemaining(self):
    """
    Returns the number of unread/unconsumed bits.
    """
    return self.NumBits() - (8*self.idx_byte + self.idx_boff) - 1

  def AllConsumed(self):
    """ Returns true if all stored bits were consumed, else returns false"""
    return self.NumBits() <= (8*self.idx_byte + self.idx_boff)

  def GetBits8(self):
    """
    Gets the next 8 unconsumed bits from the BitBucket and returns that as
    an int
    """
    raw_data = self.GetBits(8)[0]
    arg = "%c%c%c%c" % (0,0, 0, raw_data[0])
    return struct.unpack(">L", arg)[0]

  def GetBits16(self):
    """
    Gets the next 16 unconsumed bits from the BitBucket and returns that as an
    int
    """
    raw_data = self.GetBits(16)[0]
    arg = "%c%c%c%c" % (0,0, raw_data[0], raw_data[1])
    return struct.unpack(">L", arg)[0]

  def GetBits32(self):
    """
    Gets the next 32 unconsumed bits from the BitBucket and returns that as an
    int
    """
    raw_data = self.GetBits(32)[0]
    arg = "%c%c%c%c" % (raw_data[0], raw_data[1], raw_data[2], raw_data[3])
    return struct.unpack(">L", arg)[0]

  def GetBits(self, num_bits):
    """
    Gets the specified number of unconsumed bits and returns it as a list of
    ints (all of which are < 256)
    """
    old_idx_boff = self.idx_boff

    bits_available = self.NumBits() - (8*self.idx_byte + self.idx_boff)
    if num_bits > bits_available:
      print "num_bits: %d but bits_available: %d" % (num_bits, bits_available)
      raise StandardError()
    retval = []
    bits_left = num_bits
    if self.idx_boff == 0:
      while bits_left >= 8:
        retval.append(self.output[self.idx_byte])
        self.idx_byte += 1
        bits_left -= 8
      if bits_left:
        retval.append( ~(255 >> bits_left) & self.output[self.idx_byte])
        self.idx_boff += bits_left
        self.idx_boff %= 8
        bits_left = 0
    else:
      # We know there is a non-zero bit offset if we're below here.
      cur_byte = 0
      cur_boff = 0
      lob = len(self.output)
      while bits_left > 0:
        if bits_left >= 8 and lob > self.idx_byte:
          cur_byte =  255 & (self.output[self.idx_byte] << self.idx_boff)
          self.idx_byte += 1
          cur_byte |=  (self.output[self.idx_byte] >> (8 - self.idx_boff))
          retval.append(cur_byte)
          cur_byte = 0
          bits_left -= 8
        else:
          bits_to_consume = min(min(8 - cur_boff, 8 - self.idx_boff),
                                bits_left)

          c = self.output[self.idx_byte]
          c <<= self.idx_boff
          c &= 255
          cur_byte |= (c & ~(255 >> (bits_to_consume))) >> cur_boff
          bits_left -= bits_to_consume
          cur_boff += bits_to_consume
          self.idx_boff += bits_to_consume
          if cur_boff >= 8:
            retval.append(cur_byte)
            cur_byte = 0
            cur_boff -= 8
          if self.idx_boff >= 8:
            self.idx_byte += 1
            self.idx_boff -= 8
            if self.idx_boff >= 8:
              raise StandardError()
      if cur_boff:
        retval.append(cur_byte)
    if (old_idx_boff + num_bits) % 8 != self.idx_boff:
      print "old_idx_boff(%d) + num_bits(%d) != self.idx_boff(%d) " % (
          old_idx_boff, num_bits, self.idx_boff)
      print "retval: ", (retval, num_bits)
      raise StandardError()
    return (retval, num_bits)

  def DebugFormat(self):
    """
    Prints out (to stdout) a representation intended to help with debugging
    """
    print FormatAsBits((self.output, self.out_boff))
    for i in xrange(self.idx_byte*8 + self.idx_boff - 1):
      if not i % 8:
        sys.stdout.write("|")
      sys.stdout.write("-")
    print "^"

  def __repr__(self):
    return FormatAsBits((self.output, self.out_boff))

File: example_code/bit_bucket.cc
#include <sstream>
#include <string>
#include <stdlib.h>
#include <vector>
#include <iostream>
#include <algorithm>

using std::stringstream;
using std::string;
using std::vector;
using std::cerr;
using std::cout;
using std::min;
using std::ostream;
using std::flush;
using std::min;
using std::max;

template <typename T>
string FormatAsBits(const T& v, int num_bits, int offset = 0) {
  stringstream retval;
  for (int i = 0; i < num_bits; ++i) {
    int byte_idx = i / 8;
    unsigned int c = v[byte_idx];
    if ((i + offset) % 8 == 0)
      retval << "|";
    retval << ((c & (0x80U >> (i % 8))) > 0);
  }
  return retval.str();
}

template <>
string FormatAsBits<uint32_t>(const uint32_t& v, int num_bits, int offset) {
  stringstream retval;
  for (int i = 0; i < num_bits; ++i) {
    if ((i + offset) % 8 == 0)
      retval << "|";
    retval << (((v >> (31 - i)) & 0x1U) > 0);
  }
  return retval.str();
}

template <>
string FormatAsBits<uint16_t>(const uint16_t& v, int num_bits, int offset) {
  stringstream retval;
  for (int i = 0; i < num_bits; ++i) {
    if ((i + offset) % 8 == 0)
      retval << "|";
    retval << (((v >> (15 - i)) & 0x1U) > 0);
  }
  return retval.str();
}

template <>
string FormatAsBits<uint8_t>(const uint8_t& v, int num_bits, int offset) {
  stringstream retval;
  for (int i = 0; i < num_bits; ++i) {
    if ((i + offset) % 8 == 0)
      retval << "|";
    retval << (((v >> (7 - i)) & 0x1U) > 0);
  }
  return retval.str();
}

class BitBucket {
  vector<unsigned char> bsa;
  int bsa_boff;
  int idx_byte;
  int idx_boff;
  int num_bits;
  BitBucket(const BitBucket&);
 public:
  BitBucket() {
    Clear();
  }
  void Clear() {
    bsa.clear();
    bsa_boff = 0;
    idx_byte = 0;
    idx_boff = 0;
    num_bits = 0;
  }

  void Seek(unsigned int bit_loc) {
    idx_boff = bit_loc % 8;
    idx_byte = bit_loc / 8;
  }

  void SeekDelta(int bit_loc_delta) {
    Seek(min(max(0,
                 BitsConsumed() + bit_loc_delta),
             num_bits));
  }

  void StoreBit(bool bit) {
    ++num_bits;
    if (((num_bits + 7) / 8) > bsa.size())
      bsa.push_back(0);
    bsa.back() |= bit << (7 - bsa_boff);
    ++bsa_boff;
    bsa_boff %= 8;
  }

  void StoreBits(const vector<char>& inp_bytes, int inp_bits) {
    if (inp_bits == 0) return;
    int old_bsa_boff = bsa_boff;
    num_bits += inp_bits;
    if (bsa_boff == 0) {
      bsa.insert(bsa.end(), inp_bytes.begin(), inp_bytes.end());
      bsa_boff = inp_bits % 8;
      if (bsa_boff) {
        bsa.back() &= ~(0xff >> bsa_boff);  // zero out trailing  right-bits
      }
    } else {
      int leftover_bits = 0;
      if (inp_bits % 8) {
        leftover_bits = inp_bits % 8;
      } else {
        leftover_bits = 8;
      }
      int bits_left_in_byte = 8 - bsa_boff;
      bsa.reserve(bsa.size() + inp_bytes.size());
      for (int i = 0; i < inp_bytes.size(); ++i) {
        unsigned char c = inp_bytes[i];
        bsa.back() |= c >> bsa_boff;
        bsa.push_back(c << bits_left_in_byte);
      }
      if (bsa_boff + leftover_bits <= 8) {
        bsa.pop_back();
      }
      bsa_boff = (bsa_boff + leftover_bits) % 8;
      if (bsa_boff)
        bsa.back() &= ~(0xff >> bsa_boff);
    }
    if (bsa_boff != (old_bsa_boff + inp_bits) % 8) {
      cerr << "There was some logic error in the code doing StoreBits\n";
      cerr << "bsa_boff(" << bsa_boff << ") should equal: "
           << (old_bsa_boff + inp_bits) % 8 << "\n";
      abort();
    }
  }

  int NumBits() const {
    return num_bits;
  }
  int BitsConsumed() const {
    return idx_byte * 8 + idx_boff;
  }
  bool AllConsumed() const {
    return BitsConsumed() >= NumBits();
  }
  int BitsRemaining() const {
    return NumBits() - BitsConsumed();
  }

  bool GetBit() {
    bool bit = bsa[idx_byte] & (0x80 >> idx_boff);
    ++idx_boff;
    if (idx_boff >= 8) {
      ++idx_byte;
      idx_boff -= 8;
    }
    return bit;
  }

  void GetBits(vector<char>* retval, int num_bits) {
    int output_bytes = (num_bits + 7) / 8;
    retval->reserve(output_bytes);
    int old_idx_boff = idx_boff;
    if (num_bits > NumBits()) {
      cerr << "Oops, we're asking to get more bits than are available.\n";
      cerr << "Bits available: " << BitsRemaining() << "\n";
      cerr << "Bits requested: " << num_bits << "\n";
      abort();
    }
    int bits_left = num_bits;
    if (idx_boff == 0) {
      retval->insert(retval->end(), bsa.begin() + idx_byte, bsa.begin() + idx_byte + output_bytes);
      idx_byte += num_bits / 8;
      idx_boff = num_bits % 8;
      if (idx_boff) {
        retval->back() &= ~(0xff >> idx_boff);
      }
    } else { // idx_boff != 0. There WILL be shifting about.
      int idx_leftover = 8 - idx_boff;
      while (bits_left >= 8) {
        unsigned int c = bsa[idx_byte] << idx_boff;
        ++idx_byte;
        c |= bsa[idx_byte] >> idx_leftover;
        // cout << "ONE_BYTE DOWN, BITS_LEFT:" << bits_left
        //      << " " << FormatAsBits(&c, 8) << "\n";
        retval->push_back((char)c);
        bits_left -= 8;
      }
      if (bits_left) {
        //cout << "BITS LEFT: " << bits_left << "\n";
        int cur_boff = 0;
        unsigned int cur_byte = 0;
        while (true) {
          int bits_to_consume = min(min(8 - cur_boff, idx_leftover), bits_left);
          unsigned int mask = ~(0xff >> bits_to_consume);
          cur_byte |= ((bsa[idx_byte] << idx_boff) & mask) >> cur_boff;
          bits_left -= bits_to_consume;
          idx_boff += bits_to_consume;
          if (idx_boff >= 8) {
            ++idx_byte;
            idx_boff -= 8;
          }
          cur_boff += bits_to_consume;
          if (cur_boff >= 8) {
            // something is wrong.
            cerr << "Logic error. cur_boff >= 8\n";
            cerr << "cur_boff: " << cur_boff << "\n";
            cerr << "bits_left: " << bits_left << "\n";
            abort();
          }
          if (bits_left == 0) {
            // cout << "BITS LEFT: " << bits_left
            //      << " " << FormatAsBits(&cur_byte, 8) << "\n";
            retval->push_back((char)cur_byte);
            break;
          }
        }
      }
    }
    if (idx_boff != (old_idx_boff + num_bits) % 8) {
      cerr<< "idx_boff != (old_idx_boff + num_bits) % 8\n";
      cerr << "old_idx_boff: " << old_idx_boff << "\n";
      cerr << "    idx_boff: " << idx_boff << "\n";
      cerr << "    num_bits: " << num_bits << "\n";
      abort();
    }
  }

  string AsString() const {
    stringstream ss;
    ss << *this;
    return ss.str();
  }
  string DebugStr() const {
    stringstream ss;
    ss << FormatAsBits(bsa, bsa.size()*8)
       << "[" << num_bits << "," << bsa_boff << "]";
    return ss.str();
  }

  friend ostream& operator<<(ostream& os, const BitBucket& bb) {
    os << FormatAsBits(bb.bsa, bb.NumBits())
       << " [" << bb.NumBits() << "," << bb.bsa_boff << "]";
    return os;
  }
};


